nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -2007,6 +2057,265 @@ impl DataStore {
+                        Err(e) => {
+                            return Err(err.bail(
+                                VolumeReplaceSnapshotError::SnapshotReplacementError(e)
+                            ));
+                        },
+                    };
+
+                    if !old_target_in_vcr && new_target_in_vcr {
+                        // It does seem like the replacement happened
+                        return Ok(());
+                    }
+
+                    // Update the existing volume's construction request to
+                    // replace the existing target's SocketAddrV6 with the
+                    // replacement target's
+
+                    // Copy the old volume's VCR, changing out the old target
+                    // for the new.
+                    let (new_vcr, replacements) = match replace_read_only_target_in_vcr(
+                        &old_vcr,
+                        existing,
+                        replacement,
+                    ) {
+                        Ok(new_vcr) => new_vcr,
+                        Err(e) => {
+                            return Err(err.bail(
+                                VolumeReplaceSnapshotError::SnapshotReplacementError(e)
+                            ));
+                        }
+                    };
+
+                    // Expect that this only happened once. If it happened
+                    // multiple times, question everything: how would a snapshot
+                    // be used twice?!
+
+                    if replacements != 1 {
+                        return Err(err.bail(
+                            VolumeReplaceSnapshotError::UnexpectedReplacedTargets(
+                                replacements, 1,
+                            )
+                        ));
+                    }
+
+                    let new_volume_data = serde_json::to_string(
+                        &new_vcr,
+                    )
+                    .map_err(|e| {
+                        err.bail(VolumeReplaceSnapshotError::SerdeError(e))
+                    })?;
+
+                    // Update the existing volume's data
+                    diesel::update(volume_dsl::volume)
+                        .filter(volume_dsl::id.eq(volume_id))
+                        .set(volume_dsl::data.eq(new_volume_data))
+                        .execute_async(&conn)
+                        .await
+                        .map_err(|e| {
+                            err.bail_retryable_or_else(e, |e| {
+                                VolumeReplaceSnapshotError::Public(
+                                    public_error_from_diesel(
+                                        e,
+                                        ErrorHandler::Server,
+                                    )
+                                )
+                            })
+                        })?;
+
+                    // Make a new VCR that will stash the target to delete. The
```

Comment:
I know from the last round of reviews for region replacement that these dummy volumes are needed for deletes. It still kind of irks me though, and it *feels* more complicated than it needs to be. It seems to add steps to multiple sagas and extra records. I'm definitely not saying we should do anything about it in this PR, but it may be worth thinking about how we can avoid dummy volumes. It's also possible I already missed the discussion where it was decided this was the optimal way to proceed 😸

Reply:
You haven't missed any such discussion haha - I'd like to _not_ do this, but it's especially important for these read-only targets as the number of references to a region snapshot is tracked in the `volume_references` column. Simply moving the target from one volume to another in this transaction does _not_ alter the number of volume references for a region snapshot, maintaining the accounting.

Reply:
I figured it was necessary. This all makes sense. Thanks for the explanation.